staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/equality.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fieldmanager
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+var equalities = func() conversion.Equalities {
+	// Clone equality.Semantic
+	// Add a comparator for managed fields in addition to semantic equality
+	e := equality.Semantic.Copy()
+
+	err := e.AddFunc(
+		func(a, b metav1.ManagedFieldsEntry) bool {
+			// Two objects' managed fields are equivalent if, ignoring timestamp,
+			//	the objects are deeply equal.
+			a.Time = nil
+			b.Time = nil
+
+			// Purposefully using reflect.DeepEqual, not semantic equality.
+			return reflect.DeepEqual(a, b)
+		},
+	)
+
+	if err != nil {
+		panic(fmt.Errorf("failed to add ManagedFieldsEntry comparator to equalities map: %v", err))
+	}
+
+	return e
+}()
+
+// IgnoreManagedFieldsTimestampsTransformer reverts any timestamp updates
+func IgnoreManagedFieldsTimestampsTransformer(
+	_ context.Context,
+	newObj runtime.Object,
+	oldObj runtime.Object,
+) (runtime.Object, error) {
+	if equalities.DeepEqual(newObj, oldObj) {
```

Comment:
does this mean that an array field changing from empty to undefined or back is not considered a change? @apelisse will remember this discussion. This needs a very careful analysis here.

Reply:
Yeah, it's probably fine to do a non-semantic equality here, merely ignoring the timestamp. That'll also remove the need for the copy on `Equalities`.

Reply:
Actually, the way `Equalities` work, we HAVE to use the semantic comparison (since it avoids the panic on the unexported fields). On the other hand, we're fine with the managedFields not being updated in that scenario since we already know that it means there are no changes to ownership.

Reply:
The newest revision uses a version of deep equality which does not implicitly equate nil with empty slices/maps so I believe this conversation is resolved.